library/alloc/tests/vec.rs

```diff
@@ -1050,6 +1050,42 @@ fn test_from_iter_specialization_panic_drop() {
     );
 }
 
+#[test]
+fn test_from_iter_specialization_panic_during_drop_leaks() {
+    static mut DROP_COUNTER: usize = 0;
+
+    #[derive(Debug)]
+    enum Droppable {
+        DroppedTwice(Box<i32>),
+        PanicOnDrop,
+    }
+
+    impl Drop for Droppable {
+        fn drop(&mut self) {
+            match self {
+                Droppable::DroppedTwice(_) => {
+                    unsafe {
+                        DROP_COUNTER += 1;
+                    }
+                    println!("Dropping!")
+                }
+                Droppable::PanicOnDrop => {
+                    if !std::thread::panicking() {
+                        panic!();
+                    }
+                }
+            }
+        }
+    }
+
+    let _ = std::panic::catch_unwind(AssertUnwindSafe(|| {
+        let v = vec![Droppable::DroppedTwice(Box::new(123)), Droppable::PanicOnDrop];
+        let _ = v.into_iter().take(0).collect::<Vec<_>>();
+    }));
+
+    assert_eq!(unsafe { DROP_COUNTER }, 1);
```

Comment:
Does this test ensure to eventually actually free the memory that was "leaked" as part of the "panic during drop"?

If not, this would be a problem for [running the test in Miri](https://github.com/RalfJung/miri-test-libstd) -- to ensure that the standard library is memory-leak-free, Miri errors. That's why other tests like this that already exist explicit free the "leaked" memory after the test is done.

Reply:
No, it doesn't free. Can you point to a test that does the explicit free so I can steal from that?

Reply:
I think most tests just avoid having a `Box`, so even if desturctors don't get called, there's no memory leak -- the `DROP_COUNTER` is usually enough to check that everything is all right.

I thought there were also tests that used manual "late" deallocation of the `Box`, but cannot find any... do you think checking `DROP_COUNTER` is sufficient?

Reply:
That should be possible, yeah. I took the reproducer from #83629 which relied on a double-free to trigger malloc errors  and converted it into a test-case.

Reply:
Actually, that won't do since the vec's storage itself will also get leaked.

Reply:
Ah, good point!